llvm/docs/DirectX/DXILOpTableGenDesign.rst

```diff
@@ -0,0 +1,179 @@
+   by inheriting from the class
+
+   .. code-block::
+
+        // Abstract class to represent major and minor version values
+        class Version<int major, int minor> {
+          int Major = major;
+          int Minor = minor;
+        }
+
+7. Shader stages for which the operation is applicable are represented as a list of
+   concrete records that inherit from the Class
+
+   .. code-block::
+
+      class ShaderStage;
+
+
+8. All remaining properties of the operation are represented as a list as of concrete records
+   that inherit from the class
+
+   .. code-block::
+
+      class OpAttributes;
+
+   - memory access - ``ReadNone``, ``ReadNone``
+   - ``IsDerivative``, ``IsGradient``, ``IsFeedback``, ``IsWave``, ``NeedsUniformInputs``, ``IsBarrier``
+
+9. A documentation string for the operation.
+
+
+A DXIL Operation is represented by the following TableGne class by encapsulating the various
+TableGen representations of its properties described above.
+
+.. code-block::
+
+  // Abstraction DXIL Operation
+  class DXILOpPropertiesBase {
+    int OpCode = 0;                           // Opcode of DXIL Operation
+    DXILOpClass OpClass = UnknownOpClass;     // Class of DXIL Operation.
+    Intrinsic LLVMIntrinsic = ?;              // LLVM Intrinsic DXIL Operation maps to
+    list<LLVMType> OpOverloadTypes = ?; // Valid overload type
+                                              // of DXIL Operation
+    Version SMVer = ?;                        // Min Shader Model version
+    Version SMVerLinker = ?;                  // Min Shader Model required for linking
+    Version DXILVer = ?;                      // Min DXIL version
+    list<ShaderStage> ShaderStages = ?;       // List of applicable shader stages
+    list<OpAttributes> OpAttribs = ?;         // Operation attributes
+    string Doc = "";                          // A short description of the operation
+  }
+
+
+The following convenience class is used to demonstrate the definitions of a couple of
+operations:
+
+  .. code-block::
+
+      class DXILOpProperties<int opCode,
+                    Intrinsic intrinsic,
+                    list<LLVMType> overloadTypes,
+                    string doc> : DXILOpPropertiesBase {
+        int OpCode = opCode;
+        Intrinsic LLVMIntrinsic = intrinsic;
+        list<LLVMType> OpOverloadTypes = overloadTypes;
+        string Doc = doc;
+      }
+
+Additionally, following definition of ``unary`` class is also used:
+
+.. code-block::
+
+   def unary : DXILOpClass<[llvm_any_ty, LLVMMatchType<0>]>;
+
+Following is the definition of ``Sin`` and ``Cos``
+
+.. code-block::
+
+   let OpClass = unary in {
+     def Cos  : DXILOpProperties<12, int_cos, [llvm_half_ty, llvm_float_ty],
+                                "Returns cosine(theta) for theta in radians.">;
+     def Sin  : DXILOpProperties<13, int_sin, [llvm_half_ty, llvm_float_ty],
+                           "Returns sine(theta) for theta in radians.">;
```

Comment:
I'm curious what the representation looks like for a more exotic DXIL Op. Like maybe `ddx`? How do we represent that `ddx` is available in SM 2.0+ for Pixel shaders and SM 6.6+ for Compute, Mesh and Amplification?

How do we handle `RawBufferLoad` where the `double` and `i64` overloads are only valid for SM 6.3+?

Reply:
> I'm curious what the representation looks like for a more exotic DXIL Op. Like maybe `ddx`? How do we represent that `ddx` is available in SM 2.0+ for Pixel shaders and SM 6.6+ for Compute, Mesh and Amplification?
> 
> How do we handle `RawBufferLoad` where the `double` and `i64` overloads are only valid for SM 6.3+?

Thanks for flagging these. I've noted this and plan to update the design spec in subsequent PRs, if needed, as I make progress with implementation and learn about the specific requirements to represent DXIL ops that may not be sufficiently represented using the current specification mechanism.